app/src/main/java/org/mozilla/fenix/home/recentvisits/RecentVisitsFeature.kt

```diff
@@ -0,0 +1,258 @@
+     */
+    @VisibleForTesting
+    internal fun getCombinedHistory(
+        historyHighlights: List<FeatureHistoryHighlight>,
+        historyGroups: List<FeatureHistoryGroup>
+    ): List<RecentlyVisitedItem> {
+        val totalItemsCount = historyHighlights.size + historyGroups.size
+
+        return if (totalItemsCount <= MAX_RESULTS_TOTAL) {
+            getSortedHistory(
+                historyHighlights.filterByMetadata(historyGroups),
+                historyGroups
+            )
+        } else {
+            var groupsCount = 0
+            var highlightCount = 0
+            while ((highlightCount + groupsCount) < MAX_RESULTS_TOTAL) {
+                if ((highlightCount + groupsCount) < MAX_RESULTS_TOTAL &&
+                    historyHighlights.getOrNull(highlightCount) != null
+                ) {
+                    highlightCount += 1
+                }
+
+                if ((highlightCount + groupsCount) < MAX_RESULTS_TOTAL &&
+                    historyGroups.getOrNull(groupsCount) != null
+                ) {
+                    groupsCount += 1
+                }
+            }
+
+            getSortedHistory(
+                historyHighlights
+                    .filterByMetadata(historyGroups)
+                    .sortedByDescending { it.lastAccessedTime }
+                    .take(highlightCount),
+                historyGroups
+                    .sortedByDescending { it.lastAccessedTime }
+                    .take(groupsCount)
+            )
+        }
+    }
+
+    @VisibleForTesting
+    internal fun getSortedHistory(
+        historyHighlights: List<FeatureHistoryHighlight>,
+        historyGroups: List<FeatureHistoryGroup>
+    ): List<RecentlyVisitedItem> {
+        return (historyHighlights + historyGroups)
+            .sortedByDescending { it.lastAccessedTime }
+            .map { when(it) {
+                is FeatureHistoryHighlight -> RecentHistoryHighlight(
+                    title = it.historyHighlight.title ?: it.historyHighlight.url,
+                    url = it.historyHighlight.url
+                )
+                is FeatureHistoryGroup -> RecentHistoryGroup(
+                    title = it.groupName,
+                    historyMetadata = it.groupItems
+                )
+            } }
+    }
+
+    /**
+     * Filter out highlights that are already part of a history group.
+     */
+    @VisibleForTesting
+    internal fun List<FeatureHistoryHighlight>.filterByMetadata(
+        historyMetadata: List<FeatureHistoryGroup>
+    ): List<FeatureHistoryHighlight> {
+        return filterNot { highlight ->
+            historyMetadata.any {
+                it.groupItems.any {
+                    it.key.url == highlight.historyHighlight.url
+                }
+            }
+        }
+    }
+
+    override fun stop() {
+        job?.cancel()
+    }
+}
+
+@VisibleForTesting
+internal sealed class FeatureHistoryItems {
```

Comment:
nit: The `Feature` prefix confused me a bit reading through this. How about being more feature-specific

`RecentlyVisitedItem`
-> `HistoryHighlight`
->  `HistorySearchGroup`

Reply:
Ah I see, you wanted to use different names here to not collide with the types you add to the store. Hm. Makes me wonder could you use the same types for this and add `lastAccessTime` there in RecentlyVisitedItem.kt?

Otherwise, if you want to keep the types separate, maybe:
`RecentlyVisitedItemInternal`
-> `HistoryHighlight`
-> ` HistorySearchGroup`

Reply:
It would've been easier to add this property to the `RecentlyVisitedItem` but it wouldn't been used.
It's only useful in this feature, for a particular scenario - helping to keep the code a bit cleaner while data is filtered, computed, etc.
Went with the rename you proposed. Thank you!